### DISCUSSION TITLE
Make rofi more robust against broken plugins

### DISCUSSION THREAD

**Discussion by mcepl** [davatorium/rofi#1925] [(Possible) Issues]
There is now [broken rofi-calc extension](https://github.com/svenstaro/rofi-calc/issues/112), which unfortunately means that WHOLE rofi is broken, because it breaks with the error message like the one mentioned in the ticket. Which means that on my desktop which is mostly based on rofi (for the application desktop, pinentry, NetworkManager selection panel) is now completely broken.

Would it be possible to make rofi more robust against failing extensions so that only that part would be missing?

**Comment by DaveDavenport** [comment]
I don't understand what you want. Rofi correctly identified that it cannot load your plugin as it has a wrong abi.  It actually avoided causing crashes and having issues.

**Reply by mcepl** [reply]
It didn’t. Even when running `rofi` for something completely unrelated to calc, e.g., `rofi -dmenu -theme-str 'listview {lines: 0;}' -input /dev/null -password -p 'Passphrase:' -title 'Prompt for password' -mesg 'Please enter the passphrase to unlock the OpenPGP secret key:'` crashes with the same warning.

**Reply by DaveDavenport** [reply]
It should not crash, as it is a warning (as the message indicate). For me it keeps running even if it cannot load the plugin.
This is the expected behaviour.

So unless you run it with: `G_DEBUG=fatal-warnings`  set, it should not crash and does not here when I try to reproduce it.

But looking in the report: 
> When trying to run rofi-calc using the wayland fork of rofi I receive the error Rofi-WARNING **: 17:16:40.239: ABI version of plugin: 'calc.so' does not match: 00000006 expecting: 00000007.

you are not running this rofi.

**Reply by lbonn** [reply]
@mcepl yes the ABI issue is a problem with the wayland fork, I have added some details in your linked issue.

But I also could not reproduce the crash (?) even on the fork. If it's indeed crashing, you can report it to the fork's issues tracker https://github.com/lbonn/rofi/issues

**Reply by DaveDavenport** [reply]
> @mcepl yes the ABI issue is a problem with the wayland fork, I have added some details in your linked issue.

No this is not an issue or problem (with rofi or the wayland fork), It is intended behaviour because the abi in rofi for plugins changed to add new functionality.
It is completely correct for rofi to refuse to load a plugin that is compiled against another **incompatible** version.
(The warning and the refusing to load it as in the linked issue, I am unaware of other stuff the fork does).

I also do not see the linked issue talk about a crash.   I know it can be very frustrating if tools fail making your desktop not usable, but we don't know what exactly happens if not clearly told.

**Comment by mcepl** [comment]
On Wed Dec 6, 2023 at 12:07 AM CET, Dave Davenport wrote:
> No this is not an issue, It is intended behaviour because the abi in rofi for plugins changed to add new functionality.
> It is completely correct for rofi to refuse to load a plugin that is compiled against another **incompatible** version.

So, in your opinion the blame should go to the Wayland fork, which doesn’t follow new API?

Matěj

-- 
http://matej.ceplovi.cz/blog/, @***@***.***
GPG Finger: 3C76 A027 CA45 AD70 98B5  BC1D 7920 5802 880B C9D8
 
Only those laugh of wounds, who never have been hurt.
  -- despair.com

**Reply by DaveDavenport** [reply]
No, that is not what I said at all.

**Comment by mcepl** [comment]
On Tue Dec 5, 2023 at 11:52 PM CET, Dave Davenport wrote:
> But looking in the report: 
> > When trying to run rofi-calc using the wayland fork of rofi I receive the error Rofi-WARNING **: 17:16:40.239: ABI version of plugin: 'calc.so' does not match: 00000006 expecting: 00000007.
>
> you are not running this rofi.

Oh, I see, https://github.com/svenstaro/rofi-calc/issues/112#issuecomment-1841753982

Thank you for your patience.

Matěj

**Reply by DaveDavenport** [reply]
> Thank you for your patience.

In the future please try to provide more information, your original post (very understandable given your desktop not working) reads like only outing of frustration.  It is however very hard to help  if you do not provides information for us to go on.
Basic information as  exact version you are running of rofi (rofi -v)  /plugins and other relevant things as a clear description of what is happening. An empty statement like  'WHOLE rofi is broken'  does not really tell us anything.

This is also why i started my first reply with 'I don't understand' as what you asked for (keep running without the parts it cannot load) is exactly what I expect to happen given the implementation and nothing in the information you gave me suggested it didn't.
